File: back_end/test/utilities/math.test.js
const assert = require('assert');
const math = require('../../src/utilities/math');

describe('"math" utility', () => {
  describe('isNumber()', () => {
    it('identifies 7 as a number', () => {
      assert.equal(
        math.isNumber(7),
        true
      );
    });
    it('identifies 0 as a number', () => {
      assert.equal(
        math.isNumber(0),
        true
      );
    });
    it('identifies "a" as not a number', () => {
      assert.equal(
        math.isNumber('a'),
        false
      );
    });
    it('identifies NaN as not a number', () => {
      assert.equal(
        math.isNumber(NaN),
        false
      );
    });
    it('identifies Infinity as not a number', () => {
      assert.equal(
        math.isNumber(Infinity),
        false
      );
      assert.equal(
        math.isNumber(- Infinity),
        false
      );
    });
  });
  describe('remainderAndQuotient()', () => {
    it('returns an array with the remainder at index zero', () => {
      assert.equal(
        math.remainderAndQuotient(14, 3)[0],
        2
      );
      assert.equal(
        math.remainderAndQuotient(2, 3)[0],
        2
      );
      assert.equal(
        math.remainderAndQuotient(0, 3)[0],
        0
      );
    });
    it('returns an array with the quotient at index one', () => {
      assert.equal(
        math.remainderAndQuotient(14, 3)[1],
        4
      );
      assert.equal(
        math.remainderAndQuotient(2, 3)[1],
        0
      );
      assert.equal(
        math.remainderAndQuotient(0, 3)[1],
        0
      );
    });
    it('throws an exception when the denominator is zero', () => {
      assert.throws(() => {
        math.remainderAndQuotient(5, 0);
      });
      assert.throws(() => {
        math.remainderAndQuotient(0, 0);
      });
    });
  });
});

File: fastapi/tests/utilities/math_test.py
import pytest

from crossword_builder_api.utilities.math import calculate_remainder_and_quotient

def test_calculate_remainder_and_quotient():
    assert calculate_remainder_and_quotient(14, 3) == (2, 4)
    assert calculate_remainder_and_quotient(2, 3) == (2, 0)
    assert calculate_remainder_and_quotient(0, 3) == (0, 0)
    assert calculate_remainder_and_quotient(-2, 3) == (1, -1)
    assert calculate_remainder_and_quotient(-3, 3) == (0, -1)
    assert calculate_remainder_and_quotient(-4, 3) == (2, -2)
    with pytest.raises(Exception):
        calculate_remainder_and_quotient(5, -3)
    with pytest.raises(Exception):
        calculate_remainder_and_quotient(5, 0)
    with pytest.raises(Exception):
        calculate_remainder_and_quotient(0, 0)
